File: client/src/lib/permissions.ts
import { Role } from '@/types';

export type Permission =
  | 'view:dashboard'
  | 'view:anomalies'
  | 'action:mark_anomaly_reviewed'
  | 'view:predictions'
  | 'action:recalculate_prediction'
  | 'view:spare_parts'
  | 'action:approve_po'
  | 'action:manage_suppliers'
  | 'view:chat'
  | 'view:notifications'
  | 'action:trigger_notification'
  | 'manage:members'
  | 'manage:org_settings'
  | 'admin:manage_orgs'
  | 'admin:manage_users'
  | 'admin:view_tickets'
  | 'admin:platform_settings';

const ROLE_PERMISSIONS: Record<Role, Permission[]> = {
  admin: [
    'admin:manage_orgs',
    'admin:manage_users',
    'admin:view_tickets',
    'admin:platform_settings',
  ],
  manager: [
    'view:dashboard',
    'view:anomalies',
    'action:mark_anomaly_reviewed',
    'view:predictions',
    'action:recalculate_prediction',
    'view:spare_parts',
    'action:approve_po',
    'action:manage_suppliers',
    'view:chat',
    'view:notifications',
    'action:trigger_notification',
    'manage:members',
    'manage:org_settings',
  ],
  operator: [
    'view:dashboard',
    'view:anomalies',
    'view:predictions',
    'view:chat',
    'view:notifications',
  ],
  technician: [
    'view:dashboard',
    'view:anomalies',
    'action:mark_anomaly_reviewed',
    'view:predictions',
    'action:recalculate_prediction',
    'view:spare_parts',
    'view:chat',
    'view:notifications',
  ],
  procurement: [
    'view:spare_parts',
    'action:approve_po',
    'action:manage_suppliers',
    'view:notifications',
    'view:anomalies',
  ],
  viewer: [
    'view:dashboard',
    'view:anomalies',
    'view:predictions',
    'view:spare_parts',
    'view:notifications',
  ],
};

export function hasPermission(role: Role | null | undefined, permission: Permission): boolean {
  if (!role) return false;
  return ROLE_PERMISSIONS[role]?.includes(permission) ?? false;
}

export function canAccessRoute(role: Role | null | undefined, pathname: string): boolean {
  if (!role) return false;

  const routePermissions: { pattern: RegExp; permission: Permission }[] = [
    { pattern: /^\/dashboard/, permission: 'view:dashboard' },
    { pattern: /^\/anomalies/, permission: 'view:anomalies' },
    { pattern: /^\/predictions/, permission: 'view:predictions' },
    { pattern: /^\/spare-parts/, permission: 'view:spare_parts' },
    { pattern: /^\/chat/, permission: 'view:chat' },
    { pattern: /^\/notifications/, permission: 'view:notifications' },
    { pattern: /^\/members/, permission: 'manage:members' },
    { pattern: /^\/settings/, permission: 'manage:org_settings' },
  ];

  const match = routePermissions.find((r) => r.pattern.test(pathname));
  if (!match) return true;
  return hasPermission(role, match.permission);
}

export interface NavItem {
  label: string;
  href: string;
  icon: string;
  permission: Permission;
}

export const ALL_NAV_ITEMS: NavItem[] = [
  { label: 'Kontrol Paneli', href: '/dashboard', icon: 'LayoutDashboard', permission: 'view:dashboard' },
  { label: 'Anomaliler', href: '/dashboard/anomalies', icon: 'AlertTriangle', permission: 'view:anomalies' },
  { label: 'Tahminler', href: '/dashboard/predictions', icon: 'TrendingUp', permission: 'view:predictions' },
  { label: 'Yedek Parçalar', href: '/dashboard/spare-parts', icon: 'Package', permission: 'view:spare_parts' },
  { label: 'AI Asistan', href: '/dashboard/chatbot', icon: 'MessageSquare', permission: 'view:chat' },
  { label: 'Bildirimler', href: '/dashboard/notifications', icon: 'Bell', permission: 'view:notifications' },
  { label: 'Ekip', href: '/dashboard/team', icon: 'Users', permission: 'manage:members' },
  { label: 'Ayarlar', href: '/dashboard/settings', icon: 'Settings', permission: 'manage:org_settings' },
];

export function getNavItems(role: Role | null | undefined): NavItem[] {
  if (!role) return [];
  return ALL_NAV_ITEMS.filter((item) => hasPermission(role, item.permission));
}

File: server/auth/permissions.py
"""Permission system — maps roles to permissions, provides check functions."""

from enum import StrEnum

from fastapi import HTTPException, status

from auth.models import Role


class Permission(StrEnum):
    """Fine-grained permissions that are granted to roles."""

    # Dashboard
    VIEW_DASHBOARD = "view:dashboard"

    # Anomalies
    VIEW_ANOMALIES = "view:anomalies"
    MARK_ANOMALY_REVIEWED = "action:mark_anomaly_reviewed"

    # Predictions
    VIEW_PREDICTIONS = "view:predictions"
    RECALCULATE_PREDICTION = "action:recalculate_prediction"

    # Spare Parts & POs
    VIEW_SPARE_PARTS = "view:spare_parts"
    APPROVE_PO = "action:approve_po"
    MANAGE_SUPPLIERS = "action:manage_suppliers"

    # Chat
    VIEW_CHAT = "view:chat"

    # Notifications
    VIEW_NOTIFICATIONS = "view:notifications"
    TRIGGER_NOTIFICATION = "action:trigger_notification"

    # Org Management
    MANAGE_MEMBERS = "manage:members"
    MANAGE_ORG_SETTINGS = "manage:org_settings"

    # Platform Admin
    ADMIN_MANAGE_ORGS = "admin:manage_orgs"
    ADMIN_MANAGE_USERS = "admin:manage_users"
    ADMIN_VIEW_TICKETS = "admin:view_tickets"
    ADMIN_PLATFORM_SETTINGS = "admin:platform_settings"


# All org-level view permissions (used by Manager and Viewer)
_ALL_ORG_VIEW = {
    Permission.VIEW_DASHBOARD,
    Permission.VIEW_ANOMALIES,
    Permission.VIEW_PREDICTIONS,
    Permission.VIEW_SPARE_PARTS,
    Permission.VIEW_CHAT,
    Permission.VIEW_NOTIFICATIONS,
}

# All org-level action permissions (used by Manager)
_ALL_ORG_ACTIONS = {
    Permission.MARK_ANOMALY_REVIEWED,
    Permission.RECALCULATE_PREDICTION,
    Permission.APPROVE_PO,
    Permission.MANAGE_SUPPLIERS,
    Permission.TRIGGER_NOTIFICATION,
    Permission.MANAGE_MEMBERS,
    Permission.MANAGE_ORG_SETTINGS,
}

ROLE_PERMISSIONS: dict[Role, set[Permission]] = {
    Role.ADMIN: {
        Permission.ADMIN_MANAGE_ORGS,
        Permission.ADMIN_MANAGE_USERS,
        Permission.ADMIN_VIEW_TICKETS,
        Permission.ADMIN_PLATFORM_SETTINGS,
    },
    Role.MANAGER: _ALL_ORG_VIEW | _ALL_ORG_ACTIONS,
    Role.OPERATOR: {
        Permission.VIEW_DASHBOARD,
        Permission.VIEW_ANOMALIES,
        Permission.VIEW_PREDICTIONS,
        Permission.VIEW_CHAT,
        Permission.VIEW_NOTIFICATIONS,
    },
    Role.TECHNICIAN: {
        Permission.VIEW_DASHBOARD,
        Permission.VIEW_ANOMALIES,
        Permission.MARK_ANOMALY_REVIEWED,
        Permission.VIEW_PREDICTIONS,
        Permission.RECALCULATE_PREDICTION,
        Permission.VIEW_SPARE_PARTS,
        Permission.VIEW_CHAT,
        Permission.VIEW_NOTIFICATIONS,
    },
    Role.PROCUREMENT: {
        Permission.VIEW_SPARE_PARTS,
        Permission.APPROVE_PO,
        Permission.MANAGE_SUPPLIERS,
        Permission.VIEW_NOTIFICATIONS,
    },
    Role.VIEWER: _ALL_ORG_VIEW,
}


def has_permission(role: Role, permission: Permission) -> bool:
    """Check if a role has a specific permission."""
    return permission in ROLE_PERMISSIONS.get(role, set())


def check_permission(role: Role, permission: Permission) -> None:
    """Raise 403 if the role does not have the required permission."""
    if not has_permission(role, permission):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=f"Role '{role.value}' does not have permission '{permission.value}'",
        )
